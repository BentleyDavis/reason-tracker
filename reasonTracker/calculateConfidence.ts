

export interface Score {
    pctConfidence: number,
}

export interface ScoreWithDisplayData extends Score {
    pctProMeConfidence: number
    pctConMeConfidence: number
}

export interface ScoreWithParent extends Score {
    proMyParent: boolean
    pctRelevantToMyParent: number
}

export function calculateConfidence(children: ScoreWithParent[]): ScoreWithDisplayData {
    if (children.length < 1) {
        return {
            pctProMeConfidence: 1,
            pctConMeConfidence: 0,
            pctConfidence: 1,
        }
    }

    let pctProMeConfidence = 0;
    let pctConMeConfidence = 0;
    let pctConfidence = 0;

    let ChildrenWeight = 0;
    for (const child of children) {
        ChildrenWeight += weight(child);
    }

    for (const child of children) {
        const childValue =
            child.pctConfidence
            * weight(child) / ChildrenWeight // multiply by the percentage of the total children weight
            * (child.proMyParent === false ? -1 : 1); // Flip it if it is a con (not pro)

        pctConfidence += childValue

        if (childValue > 0) {
            pctProMeConfidence += childValue
        } else {
            pctConMeConfidence += -childValue
        }
    }

    return {
        pctProMeConfidence,
        pctConMeConfidence,
        pctConfidence,
    }

}

function weight(score: ScoreWithParent) {
    return Math.abs(score.pctConfidence) * score.pctRelevantToMyParent
}