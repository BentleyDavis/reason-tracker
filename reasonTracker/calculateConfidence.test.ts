import { ScoreWithParent, calculateConfidence } from "./calculateConfidence";

describe("calculateConfidence", () => {

    it("should return 0 confidence for .5 pro .5 con", () => {
        const children: ScoreWithParent[] = [
            { pctConfidence: 1, proMyParent: true, pctRelevantToMyParent: 1 },
            { pctConfidence: 1, proMyParent: false, pctRelevantToMyParent: 1 },
        ];

        const result = calculateConfidence(children);

        expect(result.pctProMeConfidence).toEqual(.5);
        expect(result.pctConMeConfidence).toEqual(.5);
        expect(result.pctConfidence).toEqual(0);
    });

    it("two half true", () => {
        const children: ScoreWithParent[] = [
            { pctConfidence: .5, proMyParent: true, pctRelevantToMyParent: 1 },
            { pctConfidence: .5, proMyParent: false, pctRelevantToMyParent: 1 },
        ];

        const result = calculateConfidence(children);

        expect(result.pctProMeConfidence).toEqual(.25);
        expect(result.pctConMeConfidence).toEqual(.25);
        expect(result.pctConfidence).toEqual(0);
    });

    it("mix 1", () => {
        const children: ScoreWithParent[] = [
            { pctConfidence: 1, proMyParent: true, pctRelevantToMyParent: 1 },
            { pctConfidence: .5, proMyParent: false, pctRelevantToMyParent: 1 },
        ];

        const result = calculateConfidence(children);
        expect(result).toMatchSnapshot()
    });

    it("should return -1 confidence for 1 con", () => {
        const children: ScoreWithParent[] = [
            { pctConfidence: 1, proMyParent: false, pctRelevantToMyParent: 1 },
        ];

        const result = calculateConfidence(children);

        expect(result.pctProMeConfidence).toEqual(0);
        expect(result.pctConMeConfidence).toEqual(1);
        expect(result.pctConfidence).toEqual(-1);
    });

    it("should handle empty children array", () => {
        const children: ScoreWithParent[] = [];

        const result = calculateConfidence(children);

        expect(result.pctProMeConfidence).toEqual(1);
        expect(result.pctConMeConfidence).toEqual(0);
        expect(result.pctConfidence).toEqual(1);
    });
});