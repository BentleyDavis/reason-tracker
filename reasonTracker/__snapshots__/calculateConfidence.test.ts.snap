// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`calculateConfidence mix 1 1`] = `
{
  "pctConMeConfidence": 0.16666666666666666,
  "pctConfidence": 0.5,
  "pctProMeConfidence": 0.6666666666666666,
}
`;
